False positive in `unconditional-recursion` rule

**Describe the bug**
The rule `unconditional-recursion` produce false positives in presence of anonymous receivers.
For example, for the following code
```go
func (*InternalMessageInfo) DiscardUnknown(m Message) {
        DiscardUnknown(m)
}
```
it will warn about an unconditional recursive call.

**To Reproduce**
Steps to reproduce the behavior:
1. I updated revive `go get -u github.com/mgechev/revive`
2. I run it with the following flags & configuration file:

```shell
revive  -config rec.toml github.com/golang/protobuf/proto/deprecated.go
```

```toml
# config file
ignoreGeneratedHeader = false
severity = "warning"
confidence = 0.5
errorCode = 0
warningCode = 0

[rule.unconditional-recursion]
```

Ouput:
```
github.com/golang/protobuf/proto/deprecated.go:92:2: unconditional recursive call
```

**Expected behavior**
No failures

--- a/rule/unconditional-recursion.go
+++ b/rule/unconditional-recursion.go
@@ -61,8 +61,10 @@ func (w lintUnconditionalRecursionRule) Visit(node ast.Node) ast.Visitor {
 	case *ast.FuncDecl:
 		var rec *ast.Ident
 		switch {
-		case n.Recv == nil || n.Recv.NumFields() < 1 || len(n.Recv.List[0].Names) < 1:
+		case n.Recv == nil:
 			rec = nil
+		case n.Recv.NumFields() < 1 || len(n.Recv.List[0].Names) < 1:
+			rec = &ast.Ident{Name: "_"}
 		default:
 			rec = n.Recv.List[0].Names[0]
 		}

--- a/testdata/unconditional-recursion.go
+++ b/testdata/unconditional-recursion.go
@@ -178,3 +178,12 @@ func urn17(ch chan int) {
 	}
 	urn17(ch) // MATCH /unconditional recursive call/
 }
+
+// Tests for #596
+func (*fooType) BarFunc() {
+	BarFunc()
+}
+
+func (_ *fooType) BazFunc() {
+	BazFunc()
+}

